lib/Dialect/TTIR/IR/TTIROps.cpp

```diff
@@ -409,34 +414,149 @@ ::mlir::LogicalResult mlir::tt::ttir::MatmulOp::verify() {
+  // Verify that the batch dimensions are broadcast compatible and construct the
+  // expected output shape
+  if (inputAShape.size() > 2 || inputBShape.size() > 2) {
+    llvm::SmallVector<int64_t> inputABatchDims, inputBBatchDims;
+
+    if (inputAShape.size() > 2) {
+      inputABatchDims.insert(inputABatchDims.begin(), inputAShape.begin(),
+                             inputAShape.end() - 2);
+    }
+
+    if (inputBShape.size() > 2) {
+      inputBBatchDims.insert(inputBBatchDims.begin(), inputBShape.begin(),
+                             inputBShape.end() - 2);
+    }
+
+    // Verify that the batch dimensions of input A and B are broadcast
+    // compatible
+    llvm::SmallVector<int64_t, 4> broadcastedShape;
+    if (!OpTrait::util::getBroadcastedShape(inputABatchDims, inputBBatchDims,
+                                            broadcastedShape)) {
+
+      // Pretty print the non-broadcastable batch dimensions
+      std::string inputABatchDimsStr = "(", inputBBatchDimsStr = "(";
+      for (size_t i = 0; i < inputABatchDims.size(); i++) {
+        inputABatchDimsStr += std::to_string(inputABatchDims[i]);
+        if (i < inputABatchDims.size() - 1) {
+          inputABatchDimsStr += ", ";
+        }
+      }
+      inputABatchDimsStr += ")";
+
+      for (size_t i = 0; i < inputBBatchDims.size(); i++) {
+        inputBBatchDimsStr += std::to_string(inputBBatchDims[i]);
+        if (i < inputBBatchDims.size() - 1) {
+          inputBBatchDimsStr += ", ";
+        }
+      }
+      inputBBatchDimsStr += ")";
+
+      return emitOpError("Batch dimensions of input A" + inputABatchDimsStr +
+                         " and B" + inputBBatchDimsStr +
+                         " are not broadcast compatible");
+    }
+
+    // Insert the broadcasted batch dimensions in the expected output shape
+    expectedOutputShape.insert(expectedOutputShape.begin(),
+                               broadcastedShape.begin(),
+                               broadcastedShape.end());
   }
-  if (outputShape[outputShape.size() - 1] !=
-      inputBShape[inputBShape.size() - 1]) {
-    return emitOpError(
-        "Output must have the same number of columns as input B");
+
+  // Insert the input A and B inner dimensions in expected output shape
+  // Consider the case where input A and B are vectors. In that case,
+  // the dimension 1 is ommited from the output shape.
+  if (inputAType.getRank() > 1) {
+    expectedOutputShape.push_back(inputAShape[inputAShape.size() - 2]);
+  }
+
+  if (inputBType.getRank() > 1) {
+    expectedOutputShape.push_back(inputBShape[inputBShape.size() - 1]);
   }
+
+  // Check the case of a vector-vector product. At this moment we don't support
+  // scalars in IR, hence check that the output is at least 1D tensor of size 1.
+  if (expectedOutputShape.size() == 0) {
+    if (outputType.getRank() < 1) {
+      return emitOpError("Scalar output is not supported, output must be at "
```

Comment:
In the past I was opposed to having scalar as a part of IR(like we can pack scalars in one element array), maybe we should reconsider that now? :)

Reply:
Yeah it's kind of a PITA, maybe we can relax and support 0 rank tensors, it's another special case we need to handle when constructing one and passing into the runtime, but maybe it's not a big deal.

Reply:
Created the following issue to track this:
https://github.com/tenstorrent/tt-mlir/issues/767